tools/lib/cmdline/cmdline.go

```diff
@@ -0,0 +1,127 @@
+	// "auto" behavior prefers binary.
+	Filey
+)
+
+func sizedBytes(flag, value string, byteSize int, decode func(string) ([]byte, error)) ([]byte, error) {
+	bytes, err := decode(value)
+	if err != nil {
+		return nil, fmt.Errorf("%s=%s could not be decoded: %v", flag, value, err)
+	}
+	if len(bytes) > byteSize {
+		return nil, fmt.Errorf("%s=%s (%v) is not representable in %d bytes", flag, value, bytes, byteSize)
+	}
+	sized := make([]byte, byteSize)
+	copy(sized, bytes)
+	return sized, nil
+}
+
+func parseBytesFromString(name string, byteSize int, in string, inform string) ([]byte, error) {
+	if !utf8.ValidString(in) {
+		return nil, fmt.Errorf("could not decode %s contents as a UTF-8 string. Try -inform=bin", name)
+	}
+	// Strict forms first.
+	if inform == "hex" {
+		return sizedBytes(name, in, byteSize, hex.DecodeString)
+	}
+	if inform == "base64" {
+		return sizedBytes(name, in, byteSize, base64.StdEncoding.DecodeString)
+	}
+	if inform != "auto" {
+		return nil, fmt.Errorf("unknown -inform=%s", inform)
+	}
+	// "auto" means to try hex encoding first, then base64.
+	if b, err := sizedBytes(name, in, byteSize, hex.DecodeString); err == nil {
+		return b, nil
+	}
+	return sizedBytes(name, in, byteSize, base64.StdEncoding.DecodeString)
+}
+
+func isBinForm(inform string, intype InputType) bool {
+	if inform == "bin" {
+		return true
+	}
+	return (intype == Filey && inform == "auto")
+}
+
+// ParseBytes returns the denoted bytes from the reader `in` or an error.
+func ParseBytes(name string, byteSize int, in io.Reader, inform string, intype InputType) ([]byte, error) {
+	inbytes, err := io.ReadAll(in)
+	if err != nil {
+		return nil, err
+	}
+	// Empty input is treated as an empty array, not a zero-filled byteSize array.
+	// This allows initial values of nil to be distinguishable from 0.
+	if len(inbytes) == 0 {
+		return nil, nil
+	}
+	if isBinForm(inform, intype) {
+		if len(inbytes) != byteSize {
+			return nil, fmt.Errorf("binary input type had %d bytes. Expect exactly %d bytes",
+				len(inbytes), byteSize)
+		}
+		return inbytes, nil
+	}
+	return parseBytesFromString(name, byteSize, strings.TrimSpace(string(inbytes)), inform)
+}
+
+// Bytes is a flag.Func parsing function that translates a string into
+// a specific byte-width array.
+//
+// A byte string can be represented as
+// *  hexadecimal encoded string if -inform=hex or -inform=auto.
+// *  base64 if -inform=base64 or -inform=auto
+//
+// Hex string decoding is attempted first with auto. The base64 encoding grammar
+// intersects with the hex encoding grammar, so -inform=auto can misbehave.
+func Bytes(name string, byteSize int, in *string) *[]byte {
+	result := &[]byte{}
+	allFlags = append(allFlags, func(inform string) error {
+		// No input means to keep the initial value.
+		if *in == "" {
+			return nil
+		}
+		bytes, err := ParseBytes(name, byteSize, strings.NewReader(*in), inform, Stringy)
+		if err != nil {
+			return err
+		}
+		*result = bytes
+		return nil
```

Comment:
I feel a function defined this way inside the "Bytes" function is less readable. Can this be a separate method and get called by "Bytes"? Or at least a comment of it would be helpful.

Reply:
The function will take the same arguments as Bytes and still return a func. allFlags is an array of functions that all get run by Parse later. I don't think lifting the lambda (anonymous function) out of the body improves clarity.